lib/collection/src/shards/replica_set/update.rs

```diff
@@ -136,13 +136,44 @@ impl ShardReplicaSet {
         operation: CollectionUpdateOperations,
         wait: bool,
     ) -> CollectionResult<UpdateResult> {
+        // `ShardRepilcaSet::update_impl` is not cancel safe, so this method is not cancel safe.
+
+        // TODO: Optimize `remotes`/`local`/`clock` locking for the "happy path"?
+        //
+        // E.g., refactor `update`/`update_impl`, so that it would be possible to:
+        // - lock `remotes`, `local`, `clock` (in specified order!) on the *first* iteration of the loop
+        // - then release and lock `remotes` and `local` *only* for all next iterations
+        // - but keep initial `clock` for the whole duration of `update`
+        let mut clock = self.clock_set.lock().await.get_clock();
+
+        loop {
+            let res = self
+                .update_impl(operation.clone(), wait, &mut clock)
+                .await?;
+
+            if let Some(res) = res {
+                return Ok(res);
+            }
+
+            log::warn!("Operation {operation:?} was rejected by some node(s), retrying...");
+        }
```

Comment:
Do we still need to update our own clock to the newest we've seen?

If it's done somewhere else, I've missed it.

And obviously we want to limit retries here.

Reply:
> If it's done somewhere else, I've missed it.

It's done in `update_impl`.

Reply:
Of course, it's just our old logic of advancing with all tags we receive. Thanks!